coalib/misc/Caching.py

```diff
@@ -0,0 +1,171 @@
+            os.remove(file_path)
+
+    return ret
+
+
+def add_to_changed_files(project_dir, changed_files):
+    """
+    Saves the set of files in ``changed_files`` to disk for future use
+    in ``update_last_coala_run_time``.
+
+    :param project_dir:   The root directory of the project on which
+                          coala is to be run.
+    :param changed_files: A set of files that had changed since the last
+                          time coala was run.
+    """
+    data = pickle_load(changed_files_db, {})
+
+    if project_dir in data.keys():
+        data[project_dir].update(changed_files)
+    else:
+        data[project_dir] = changed_files
+
+    pickle_dump(changed_files_db, data)
+
+
+def get_last_coala_run_time(project_dir):
+    """
+    Get the last time coala was run on the project in epoch format
+    (number of seconds since Jan 1, 1970).
+
+    :param project_dir: The root directory of the project on which
+                        coala is to be run.
+    :return:            Returns a dict of files with file path as key
+                        and the last time that coala was run on that file
+                        as value. Returns None if this the first time coala
+                        is run on that project.
+    """
+    data = pickle_load(caching_db, {})
+    if project_dir in data.keys():
+        return data[project_dir]
+    return None
+
+
+def add_new_files_since_last_run(project_dir, last_coala_run, new_files):
+    """
+    Start tracking new files given in ``new_files`` by adding them to the
+    database.
+
+    :param project_dir:    The root directory of the project on which
+                           coala is to be run.
+    :param last_run_coala: Previously available data to which the new files
+                           need to be appended.
+    :param new_files:      The list of new files that need to be tracked.
+                           These files are initialized with their last
+                           modified tag as -1.
+    """
+    if last_coala_run is None:
+        last_coala_run = {}
+    for new_file in new_files:
+        last_coala_run[new_file] = -1
+    data = pickle_load(caching_db, {})
+    data[project_dir] = last_coala_run
+    pickle_dump(caching_db, data)
+
+
+def update_last_coala_run_time(project_dir):
+    """
+    Update the last time coala was run on the project for each file
+    to the current time.
+
+    :param project_dir: The root directory of the project on which
+                        coala is to be run.
+    """
+    current_time = calendar.timegm(time.gmtime())
+    data = pickle_load(caching_db, {})
+    changed_files = pickle_load(changed_files_db, {})
+
+    # update_last_coala_run_time is always ran after
+    # add_new_files_since_last_run, or the project has already been
+    # initialized in the caching database.
+    if project_dir not in data:
+        return
+
+    for file in data[project_dir].keys():
```

Comment:
@hypothesist IMO you should not prefer to use `file` as it's an alias of `open` method in Python 2. Perhaps in Python 3 it's no more used but it would be better not to use it. 
And btw you realise that if you want to iterate through dictionary keys then you can just do
```python
for my_key in data[project_dir]
```
rather than
```python
for my_key in data[project_dir].keys()
```